### PR TITLE
Fix read and write confilicts

### PR DESCRIPTION
**Pull request by wzshiming**
Signed-off-by: Shiming Zhang <wzshiming@foxmail.com>

Fix read and write confilicts

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[x] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Review by howardjohn — CHANGES_REQUESTED** [review]
Thanks for the PR! Please add a description of all PRs you submit. Without this, its hard to understand the motivation for the change and what its intended to do. Please let me know when you have added this and I will look at the change. Thanks!

**Comment by wzshiming** [comment]
`its.Expires` and `its.Token` it is now possible to be read and written concurrently, there is a chance of panic.
This PR optimize lock logic to ensure thread safety

**Comment by howardjohn** [comment]
Is it theoretical or has it crashed? can you show a log of a crash?

File: pkg/kube/rpc_creds.go [hunk 1]
```diff
@@ -83,10 +89,12 @@ func (its *tokenSupplier) GetRequestMetadata(ctx context.Context, uri ...string)
 				log.Infof("GetRequestMetadata failed to recreate token: %v", err.Error())
 			}
 		}
+		token = its.Token
+		its.mu.Unlock()
 	}
 
 	return map[string]string{
-		"authorization": "Bearer " + its.Token,
+		"authorization": "Bearer " + token,
```
Comment (howardjohn) [inline pkg/kube/rpc_creds.go:97]: doesn't this mean if its not going to expire token will be `""`?

**Review by howardjohn — CHANGES_REQUESTED** [review]
Can you add tests? I agree there is a race today but don't feel comfortable with the changes without tests

File: pkg/kube/rpc_creds_test.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+package kube
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+
+	"istio.io/istio/galley/pkg/testing/mock"
```
Comment (howardjohn) [inline pkg/kube/rpc_creds_test.go:23]: Please don't use this package, it is deprecated. Use `kube.NewFakeClient()`. Thanks!

File: pkg/kube/rpc_creds_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	m2, err := perCred.GetRequestMetadata(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !reflect.DeepEqual(m1, m2) {
+		t.Fatalf("Unexpectedly getting a new tokens")
+	}
+
+	time.Sleep(time.Second)
```
Comment (howardjohn) [inline pkg/kube/rpc_creds_test.go:51]: time.Sleep is not appropriate in tests, see https://github.com/istio/istio/wiki/Test-Flakes

File: pkg/kube/rpc_creds_test.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+	"istio.io/istio/pkg/test/util/retry"
+	"istio.io/istio/tests/util"
+)
+
+func Test_tokenSupplier_GetRequestMetadata(t *testing.T) {
+	ctx := context.Background()
+	cli := NewFakeClient()
+	clientset := cli.Kube().(*fake.Clientset)
+	clientset.PrependReactor("create", "serviceaccounts",
```
Comment (wzshiming) [inline pkg/kube/rpc_creds_test.go:37]: Expected here is "serviceaccounts/token", but `fake.NewSimpleClientset` does not support. kubernetes/kubernetes#100938

**Review by howardjohn — APPROVED** [review]
Thanks!

**Pull request merged**